system-addon/content-src/components/TopSites/TopSite.jsx

```diff
@@ -21,25 +21,38 @@ export class TopSiteLink extends React.PureComponent {
    */
   _allowDrop(e, index) {
     let draggedIndex = parseInt(e.dataTransfer.getData("text/topsite-index"), 10);
-    if (!isNaN(draggedIndex) && draggedIndex !== index) {
+    if (!isNaN(draggedIndex)) {
       return true;
     }
     return false;
   }
+
+  /*
+   * Helper to determine whether the drop zone should allow a dragEnter. This is to
+   * avoid entering the site that is already being dragged during "drop preview".
+   */
+  _allowEnter(e, index, url) {
+    return this._allowDrop(e, index) && (!url || e.dataTransfer.getData("text/topsite-url") !== url);
+  }
   onDragEvent(event) {
     switch (event.type) {
       case "dragstart":
         event.dataTransfer.effectAllowed = "move";
         event.dataTransfer.setData("text/topsite-index", this.props.index);
+        event.dataTransfer.setData("text/topsite-url", this.props.link.url);
```

Comment:
This data seems unnecessary. Here's the `event.dataTransfer.types` when dragging a top site:
`[ "text/x-moz-url", "text/x-moz-url-data", "text/x-moz-url-desc", "text/uri-list", "text/_moz_htmlcontext", "text/_moz_htmlinfo", "text/html", "text/plain", "text/topsite-index", "text/topsite-url" ]`

With matching data:
```
text/x-moz-url https://twitter.com/
twitter

text/x-moz-url-data https://twitter.com/

text/x-moz-url-desc twitter

text/uri-list https://twitter.com/

text/_moz_htmlcontext <html dir="ltr" lang="en-US"><body class="activity-stream"><div id="root"><div class="outer-wrapper fixed-to-top" data-reactroot="" data-reactid="1" data-react-checksum="-1133339003"><main data-reactid="2"><div class="body-wrapper on" data-reactid="11"><section class="collapsible-section top-sites animation-enabled" data-reactid="12"><div class="section-body" data-reactid="29"><ul class="top-sites-list" data-reactid="30"><li class="top-site-outer" draggable="true"><div class="top-site-inner"></div></li></ul></div></section></div></main></div></div></body></html>
text/_moz_htmlinfo 0,0

text/html <a href="https://twitter.com/"><div class="tile" aria-hidden="true" data-fallback="t"><div class="top-site-icon rich-icon" style="background-image: url(&quot;resource://activity-stream/data/content/tippytop/images/twitter-com@2x.png&quot;);"></div></div><div class="title "><span dir="auto">twitter</span></div></a>

text/plain https://twitter.com/

text/topsite-index 5

text/topsite-url https://twitter.com/
```